smart-engine/src/main/java/org/smartdata/server/engine/rule/SmallFilePlugin.java

```diff
@@ -0,0 +1,349 @@
+            SmartFilePermission filePermission = new SmartFilePermission(fileInfo);
+            if (filePermissionMap.containsKey(filePermission)) {
+              filePermissionMap.get(filePermission).add(object);
+            } else {
+              smallFileInfoMap.put(object, fileInfo);
+              List<String> list = new ArrayList<>();
+              list.add(object);
+              filePermissionMap.put(filePermission, list);
+            }
+          } else {
+            LOG.debug("Invalid file {} for small file compact.", object);
+          }
+        } catch (MetaStoreException e) {
+          LOG.error(String.format("Failed to get file info of %s.", object), e);
+        }
+      }
+
+      // Split small files according to the batch size
+      List<String> smallFileList = new ArrayList<>();
+      for (List<String> listElement : filePermissionMap.values()) {
+        int size = listElement.size();
+        for (int i = 0; i < size; i += batchSize) {
+          int toIndex = (i + batchSize <= size) ? i + batchSize : size;
+          String smallFiles = new Gson().toJson(listElement.subList(i, toIndex));
+          smallFileList.add(smallFiles);
+        }
+      }
+
+      return smallFileList;
+    } else {
+      return objects;
+    }
+  }
+
+  @Override
+  public CmdletDescriptor preSubmitCmdletDescriptor(
+      final RuleInfo ruleInfo, TranslateResult tResult, CmdletDescriptor descriptor) {
+    for (int i = 0; i < descriptor.getActionSize(); i++) {
+      if (COMPACT_ACTION_NAME.equals(descriptor.getActionName(i))) {
+        String smallFiles = descriptor.getActionArgs(i).get(HdfsAction.FILE_PATH);
+        if (smallFiles != null && !smallFiles.isEmpty()) {
+          // Check if small file list is empty
+          ArrayList<String> smallFileList = new Gson().fromJson(
+              smallFiles, new TypeToken<ArrayList<String>>() {
+              }.getType());
+          if (smallFileList == null || smallFileList.isEmpty()) {
+            continue;
+          }
+
+          // Get the first small file info
+          String firstFile = smallFileList.get(0);
+          FileInfo firstFileInfo;
+          if (smallFileInfoMap.containsKey(firstFile)) {
+            firstFileInfo = smallFileInfoMap.get(firstFile);
+          } else {
+            try {
+              firstFileInfo = metaStore.getFile(firstFile);
+              if (firstFileInfo == null) {
+                LOG.debug("{} is not exist!!!", firstFile);
+                continue;
+              }
+            } catch (MetaStoreException e) {
+              LOG.error("Failed to get file info of: " + firstFile, e);
+              continue;
+            }
+          }
+
+          // Get valid compact action arguments
+          SmartFilePermission firstFilePermission = new SmartFilePermission(
+              firstFileInfo);
+          String firstFileDir = firstFile.substring(0, firstFile.lastIndexOf("/") + 1);
```

Comment:
For two files: /dir1/dir2/a, /dir1/dir3/b,  which dir should the container file located?